builtin/providers/aws/resource_aws_s3_bucket_object.go

```diff
@@ -0,0 +1,117 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func resourceAwsS3BucketObject() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsS3BucketObjectPut,
+		Read:   resourceAwsS3BucketObjectRead,
+		Update: resourceAwsS3BucketObjectPut,
+		Delete: resourceAwsS3BucketObjectDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"source": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"etag": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsS3BucketObjectPut(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	source := d.Get("source").(string)
+
+	file, err := os.Open(source)
+
+	if err != nil {
+		return fmt.Errorf("Error opening S3 bucket object source (%s): %s", source, err)
+	}
+
+	resp, err := s3conn.PutObject(
+		&s3.PutObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(key),
+			Body:   file,
+		})
+
+	if err != nil {
+		return fmt.Errorf("Error putting object in S3 bucket (%s): %s", bucket, err)
+	}
+
+	d.Set("etag", resp.ETag)
+	d.SetId(key)
+	return nil
+}
+
+func resourceAwsS3BucketObjectRead(d *schema.ResourceData, meta interface{}) error {
+	s3conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	key := d.Get("key").(string)
+	etag := d.Get("etag").(string)
+
+	resp, err := s3conn.HeadObject(
+		&s3.HeadObjectInput{
+			Bucket:  aws.String(bucket),
+			Key:     aws.String(key),
+			IfMatch: aws.String(etag),
+		})
+
+	if err != nil {
+		// if there is an error reading the object we assume it's not there.
+		d.SetId("")
+		log.Printf("Error Reading Object (%s): %s", key, err)
+		return nil
+	}
```

Comment:
@catsby from that output it looks like something like this would work for a scope:

```go
if awsErr, ok := err.(awserr.RequestError); ok && awsErr.StatusCode == 404 {
  // ...
}
```

Similar to how we do it [for S3 Buckets](https://github.com/hashicorp/terraform/blob/29d4a3ee0a2cda946f4138b8de8c774e535d8fe8/builtin/providers/aws/resource_aws_s3_bucket.go#L165). `err.(awserr.RequestFailure)` might work too, which would make us able to use literally the same snippet.

Let me know what you think.